### DISCUSSION TITLE
JSON handling in Lua: related to post #196

### DISCUSSION THREAD

**Discussion by shooking** [RomanKubiak/ctrlr#292] [General]
Hi all
So I saw the discussion on [json Encoding of tables containing MemoryBlocks #196
](https://github.com/RomanKubiak/ctrlr/discussions/196).

### Upfront question

Is the "require" method guaranteed to work with Lua in Ctrlr or am I just lucky with this luajson on Linux?

What better ways are there for reading (and updating) cached data.

Instead of writing out JSON I could write to any format I like.
Are there better data exchange formats for Lua/Ctrlr?
The thing I like about the JSON is it can handle an array of parameters (anywhere between 4 to 9) and I can count them when readi the JSON as a table.

### Background to question

On discussion 196 the issues seems to be JSON cannot handle BINARY (nor timestamps for that matter) - but I didnt follow where JSON was used in that discussion? It seemed to be manipulating memory blocks? Did I misunderstand?

Here's what I did for handling valid JSON - I want to read in some state to a panel from file.
Just wondering if there are better ways to do this - any hints gratefully accepted.

### Motivation

I am reverse engineering Zoom pedals [ZoomPedalFun](https://github.com/shooking/ZoomPedalFun).
Between the Internet and I, we have managed to get JSON format to describe an FX say.

```
pi@raspberrypi:~/Software/zoom-zt2/G1XFOUR_ZD2 $ more COMP.ZD2.json 
{
      "FX": {
            "name": "Comp",
            "description": "This compressor in the style of the MXR Dyna Comp.",
            "version": "1.40",
            "id": 16,
            "groupid": 1
      },
      "Parameters": [
            {
                  "name": "Sense",
                  "explanation": "Adjusts the sensitivity of the effect.",
                  "blackback": false,
                  "pedal": false,
                  "mmax": 10,
                  "mcurr": 6
            },
            {
                  "name": "ATTCK",
                  "explanation": "Sets compressor attack speed to Fast or Slow."
,
                  "blackback": false,
                  "pedal": false,
                  "mmax": 1,
                  "mcurr": 0
            },
            {
                  "name": "Tone",
                  "explanation": "Adjusts the tone.",
                  "blackback": false,
                  "pedal": false,
                  "mmax": 10,
                  "mcurr": 6
            },
            {
                  "name": "VOL",
                  "explanation": "Adjusts the volume.",
                  "blackback": false,
                  "pedal": false,
                  "mmax": 100,
                  "mcurr": 80
            }
      ]
}
```

So I can plug in a pedal, get it to emit the state of that pedal (same for patches - just not shown).

I wasnt sure how to consume this in Ctrlr. I found [Luarocks and luajson](https://postintrend.com/breaking/lua-json-how-json-works-in-lua-example-of-lua-json/)

And if I install the above on the Pi, I can get something working - sure it might not work on other platforms but my aim has always been to use the Pi to control stuff.

In Ctrlr:
```
function filetest()
	-- Your method code here
lunajson = require 'lunajson'

f1=io.open("/home/pi/Software/zoom-zt2/G1XFOUR_ZD2/COMP.ZD2.json", "r")
contents = f1:read( "*all" )
xfile = {}
xfile = lunajson.decode( contents )

for k,v in pairs(xfile) do
    debugConsole("k is "..k)
    for k1, v1 in pairs(v) do
        debugConsole ("k1 is "..k1)
    end
end
debugConsole(xfile["FX"]["name"])
debugConsole(xfile["FX"]["description"])
debugConsole(xfile["FX"]["version"])
debugConsole(""..xfile["FX"]["id"])
debugConsole(""..xfile["FX"]["groupid"])

debugConsole(xfile["Parameters"][1]["name"])
debugConsole(xfile["Parameters"][1]["explanation"])
debugConsole(""..xfile["Parameters"][1]["mmax"])
debugConsole(""..xfile["Parameters"][1]["mcurr"])
local blackback=xfile["Parameters"][1]["blackback"]
local pedalCapable=xfile["Parameters"][1]["pedal"]
if blackback == true then
    debugConsole("blackback is true")
else
    debugConsole("blackback is false")
end

if pedalCapable == true then
    debugConsole("can assign to pedal")
else
    debugConsole("cannot assign to pedal")
end

io.close()
end
```
My debugConsole(string) prints a string if global debugMe is true.
The hardest part was understanding IF I read in an integer, I have to concat that to a blank string to print it - doh.

So I can read the data in (console output)

```

>>> filetest()
k is FX
k1 is name
k1 is id
k1 is description
k1 is version
k1 is groupid
k is Parameters
k1 is 1
k1 is 2
k1 is 3
k1 is 4
Comp
This compressor in the style of the MXR Dyna Comp.
1.40
16
1
Sense
Adjusts the sensitivity of the effect.
10
6
blackback is false
cannot assign to pedal
```

### Pros and Cons

I can use external Python/C++ to interrogate the pedal and create a cache of it, persisted as JSON.
Then I can use Ctrlr to read these in, write out changes (if user rename patch/values).
Just use Ctrlr as a value modifier and rendered (with touchscreen) - so I can see the pedal values when I play - eyes too old for tiny LCDs.

But a big con - is the "require" method guaranteed to work with Lua or am I just lucky?
What better ways are there for reading (and updating) cached data.
Instead of writing out JSON I could write to any format I like.
Are there better data exchange formats for Lua/Ctrlr?

**Comment by dobo365** [comment]
I'm handling all data with "sysex" files that I handle internally as memory blocks. You need only to decide on the position and meaning of each byte or pair of bytes.
Started with that as this is what was provided by the manufacturers (also using bit packing) but then used the same (without bit packing) for my own sysex fies.
JSON has the advantage it is directly readable but probably less fast.
With sysex binary files (also used for complete bank of patches), I can also perform any operation between the files, replace...
Looking at the Typhon, I learned they are using protobuf which provides the advantage of taking care of changes in the schema.
I found some Lua source code to include in my panel to treat that but didn't try it yet.

**Reply by shooking** [reply]
Well here I want to load in from a file. So I would have to encode all the values  into 7 bit form. The verbose JSON seems easier to manage. I will look into other methods you suggested- I will need to be able to write that format from Python or C++.

**Comment by Godlike-Productions** [comment]
I use raw sysex for compatible dumps, but Use JSON for internal structures, where I've added data that the synth doesn't handle, such as librarian data.

JSON is handled with the built in json4lua methods, and that;s what I've used to encode and decode JSON.  The only extra step I've had to take is to convert memory blocks to strings using toHexString and to convert them back using createFromTable.  Ctrlr handles the JSON heirarchies for your and seems to work well.

**Reply by shooking** [reply]
Built in - so I don't need to load the luarocks myself? It is already in the Lua of Ctrlr?

Do I need to require anything or it is just there?

Best regards
Steve

**Reply by shooking** [reply]
Is see now - you use json.encode and json.decode in final summary routines on original thread without defining it ....so it's built in. I will try it thanks. Looks promising.

**Reply by shooking** [reply]
so the build in json works for me thanks - I didnt know it exists.
The only odd behaviour I can see - when I load the json in from file and decode it, it seems the debugger spits out various incarnations of the string in its console.

But when I debugConsole("test")
and then test the access, that is ok.
So I will use the internal json thanks and ignore the cruft.

**Comment by dobo365** [comment]
I didn't know there was a json handling library integrated within Ctrlr.
Can you post some examples of code and some links to more info/description?

**Reply by shooking** [reply]
Example code was in original post.just use json.encode instead of the require it didand remember if it's an int you need to concat it "".. before printing :-)

**Reply by dnaldoog** [reply]
> Example code was in original post.just use json.encode instead of the require it didand remember if it's an int you need to concat it "".. before printing :-)

It looks like **int** and **boolean** don't need to be stringified! In my example though reading _boolean_ back was not possible of course when outputting table as string to console()

**Reply by Godlike-Productions** [reply]
Only  memoryblocks should need to be stringified

To output boolean to console use

console (toString(booleanVariable))

**Reply by dnaldoog** [reply]
@Godlike-Productions 
I didn't know that booleans can be printed using _tostring()_! Thanks.
I updated the decode function to make it simpler.

**Reply by dobo365** [reply]
Thx. I read it when I first saw your initial post then forgot it. Hahaha...

**Reply by shooking** [reply]
yes: JSON has string, int, boolean and base64 for binary support.
It doesnt handle date/time.
I use it a lot in my professional work (short hand XML for data exchange).

If you take the original JSON I read from a file, this is generated by Python interrogating the Zoom B1XFour - it is a  modified version of what Mungewell was doing for FX injection (I found some more ways to interpret the data and emitted as JSON).

So I added a luarocks JSON processors onto the Pi, then could get it to work.
It is cool to know there is a builtin way - so I dont have to worry about being too non-standard (there again Ctrlr on a Pi is a little nonstandard)

**Comment by dnaldoog** [comment]
Here is an adapted example (improved?) from that original post from the old website. :)


<pre>
<code>
encodeJSON = function()
   -- all a bit clunky but for demonstration purposes
    local elements = 3
    local t = {}
    for i = 1, elements do -- initialise matrix
        t[i] = {}
    end

    t[1].name = "Patch 01"
    t[1].sysex = MemoryBlock("f0 01 f7"):toHexString(1)
    t[1].switch = false
    t[1].value = 0x64

    t[2].name = "Resonance"
    t[2].sysex = MemoryBlock("f0 01 45 33 0a f7"):toHexString(1)
    t[2].switch = true
    t[2].value = 0xff

    t[3].name = "LFO Rate"
    t[3].sysex = MemoryBlock("f0 02 45 33 0a 66 77 7f f7"):toHexString(1)
    t[3].switch = false
    t[3].value = 0x7f

    local json_string = json.encode(t)
    panel:getLabel("enc"):setText(json_string) 
-- I send the output to a label in this example, but maybe could be a return value
end
</code>
</pre>


<pre>
<code>
decodeJSON = function()
    local str = L(panel:getLabel("enc"):getText())
	if String(str):isEmpty() then return end
    local obj = json.decode(str)
    local tab = {} -- initialise multidimensional array
    for i = 1, #obj do
        tab[i] = {name = nil, value = nil, switch = nil, sysex = nil}
    end
    for i, v in ipairs(obj) do
        for k, w in pairs(v) do
            tab[i][k] = w
        end
    end

-- output results
    for i, v in pairs(tab) do
        for j, w in pairs(v) do
            local output = string.format("[%d] %s=%s", i, j, tostring(w))
            -- tostring(w) allows booleans to be printed 
            console(output)
        end
        console("------------------")
    end
end
</code>
</pre>


![jsonsshot](https://user-images.githubusercontent.com/76831683/118571906-22c3f600-b7ba-11eb-89bd-7e95da3a08f0.png)

[JSON example_1_1_0.zip](https://github.com/RomanKubiak/ctrlr/files/6498252/JSON.example_1_1_0.zip)

**Comment by Godlike-Productions** [comment]
json.encode and json.decode will handle tables (and tables of tables) automagically. You can just parse the table and encode will do it's things, building a valid json file.  You don't need to manually handle the table steps except to do the type conversion

**Reply by dobo365** [reply]
Can you give an example or modify the example above? I suppose that the encode function above is OK but that you are speaking about the decode one that could be simplified. Right?

**Reply by dnaldoog** [reply]
Yes I was wondering that too. json.decode returns a table. I converted it into another table but that’s just for demonstration purposes. Maybe the suggestion is that I should have worked with the returned obj directly?

**Reply by shooking** [reply]
yes this is what I did. I had to run some lua interpreter and debug it - then once I got my head around it I could use external require 'lunajson' but then someone above replied it is built in.
```
f1=io.open("/home/pi/Software/zoom-zt2/G1XFOUR_ZD2/COMP.ZD2.json", "r")
contents = f1:read( "*all" )
-- was using a package BUT turns out json is built in.
--xfile = lunajson.decode( contents )
xfile = json.decode( contents )

debugConsole("Checking values")

debugConsole(xfile["FX"]["name"])
debugConsole(xfile["FX"]["description"])
debugConsole(xfile["FX"]["version"])
debugConsole(""..xfile["FX"]["id"])
```

So this allows me to generate the JSON, cache it, and consume it in Ctrlr.
And if I decide to use a different platform, well I can read the JSON into that and do the needful.

**Comment by dobo365** [comment]
Great info guys!
As I said above I handle everything by sysex/binary files so I created my own files to handle metadata.
Anyway, this way with json is also nice.
I will compare this with the protobuf way of doing it (this is used by the Typhon).

**Comment by bijlevel** [comment]
I'm not sure this will contribute to the discussion, but in the 'ctrlr master' I did find this 'json4lua folder:

[json4lua.zip](https://github.com/RomanKubiak/ctrlr/files/6500239/json4lua.zip)
